File: myfirstmod/src/main/java/com/example/dartbridge/proxy/ProxyRegistry.java
package com.example.dartbridge.proxy;

import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents;
import net.minecraft.core.Registry;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.Identifier;
import net.minecraft.resources.ResourceKey;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.CreativeModeTabs;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.block.state.BlockBehaviour;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;

/**
 * Registry for Dart-defined proxy blocks.
 *
 * This registry manages the lifecycle of proxy blocks and provides
 * methods for registering new blocks from Dart code.
 *
 * The registration process is two-phase:
 * 1. createBlock() - Called from Dart to reserve a handler ID and store block settings
 * 2. registerBlock() - Called from Dart to register the block with Minecraft's registry
 *
 * This two-phase approach is needed because Minecraft requires a ResourceKey during
 * block construction, but we don't know the block's namespace:path until registration.
 */
public class ProxyRegistry {
    private static final Logger LOGGER = LoggerFactory.getLogger("ProxyRegistry");
    private static final Map<Long, DartBlockProxy> blocks = new HashMap<>();
    private static final Map<Long, BlockSettings> pendingSettings = new HashMap<>();
    private static long nextHandlerId = 1;

    /**
     * Holds block settings between createBlock() and registerBlock() calls.
     */
    private record BlockSettings(float hardness, float resistance, boolean requiresTool) {}

    /**
     * Create a new DartBlockProxy with the given settings.
     * Returns the handler ID that links to this block.
     *
     * Called from Dart via JNI during the first phase of block registration.
     *
     * @param hardness Block hardness (time to break). Use -1 for unbreakable.
     * @param resistance Explosion resistance.
     * @param requiresTool Whether the block requires the correct tool to drop items.
     * @return The handler ID to use when registering the block.
     */
    public static long createBlock(float hardness, float resistance, boolean requiresTool) {
        long handlerId = nextHandlerId++;

        // Store settings for use during registerBlock()
        pendingSettings.put(handlerId, new BlockSettings(hardness, resistance, requiresTool));

        LOGGER.info("Prepared DartBlockProxy slot with handler ID: {}", handlerId);
        return handlerId;
    }

    /**
     * Register the block with Minecraft's registry.
     * Must be called during mod initialization before registry freeze.
     *
     * Called from Dart via JNI during the second phase of block registration.
     *
     * @param handlerId The handler ID returned by createBlock().
     * @param namespace The block namespace (e.g., "dartmod").
     * @param path The block path (e.g., "example_block").
     * @return true if registration succeeded, false otherwise.
     */
    public static boolean registerBlock(long handlerId, String namespace, String path) {
        BlockSettings settings = pendingSettings.get(handlerId);
        if (settings == null) {
            LOGGER.error("Cannot register block: handler ID {} not found (createBlock not called?)", handlerId);
            return false;
        }

        try {
            // Create resource keys
            ResourceKey<net.minecraft.world.level.block.Block> blockKey = ResourceKey.create(
                Registries.BLOCK,
                Identifier.fromNamespaceAndPath(namespace, path)
            );
            ResourceKey<Item> itemKey = ResourceKey.create(
                Registries.ITEM,
                Identifier.fromNamespaceAndPath(namespace, path)
            );

            // Create block properties with the settings from createBlock()
            BlockBehaviour.Properties properties = BlockBehaviour.Properties.of()
                .strength(settings.hardness(), settings.resistance())
                .setId(blockKey);

            if (settings.requiresTool()) {
                properties = properties.requiresCorrectToolForDrops();
            }

            DartBlockProxy block = new DartBlockProxy(properties, handlerId);
            blocks.put(handlerId, block);
            pendingSettings.remove(handlerId); // Clean up pending settings

            // Register the block
            Registry.register(BuiltInRegistries.BLOCK, blockKey, block);

            // Also register a BlockItem so it appears in creative inventory
            BlockItem blockItem = new BlockItem(block,
                new Item.Properties().setId(itemKey).useBlockDescriptionPrefix());
            Registry.register(BuiltInRegistries.ITEM, itemKey, blockItem);

            // Add to Building Blocks creative tab
            ItemGroupEvents.modifyEntriesEvent(CreativeModeTabs.BUILDING_BLOCKS).register(entries -> {
                entries.accept(blockItem);
            });

            LOGGER.info("Registered block: {}:{} with handler ID {}", namespace, path, handlerId);
            return true;
        } catch (Exception e) {
            LOGGER.error("Failed to register block {}:{}: {}", namespace, path, e.getMessage());
            e.printStackTrace();
            return false;
        }
    }

    /**
     * Get a block by its handler ID.
     */
    public static DartBlockProxy getBlock(long handlerId) {
        return blocks.get(handlerId);
    }

    /**
     * Get all registered handler IDs.
     */
    public static long[] getAllHandlerIds() {
        return blocks.keySet().stream().mapToLong(Long::longValue).toArray();
    }

    /**
     * Get the number of registered blocks.
     */
    public static int getBlockCount() {
        return blocks.size();
    }
}

File: dart_mc_bridge/fabric_bridge/src/main/kotlin/com/example/dartbridge/proxy/ProxyRegistry.kt
package com.example.dartbridge.proxy

import com.example.dartbridge.DartBlockProxy
import net.minecraft.block.AbstractBlock
import net.minecraft.item.BlockItem
import net.minecraft.item.Item
import net.minecraft.registry.Registries
import net.minecraft.registry.Registry
import net.minecraft.util.Identifier

/**
 * Registry for Dart-defined proxy blocks.
 *
 * This registry manages the lifecycle of proxy blocks and provides
 * methods for registering new blocks from Dart code via JNI.
 *
 * Each block is assigned a unique handler ID that links it to its
 * corresponding Dart-side CustomBlock instance.
 */
object ProxyRegistry {
    private val blocks = mutableMapOf<Long, DartBlockProxy>()
    private var nextHandlerId = 1L

    /**
     * Create a new DartBlockProxy with the given settings.
     * Returns the handler ID that links to this block.
     *
     * Called from Dart via JNI.
     *
     * @param hardness How long it takes to break the block (e.g., stone = 1.5)
     * @param resistance How resistant the block is to explosions (e.g., stone = 6.0)
     * @param requiresTool Whether the block requires a tool to drop items
     * @return The handler ID for this block
     */
    @JvmStatic
    fun createBlock(
        hardness: Float,
        resistance: Float,
        requiresTool: Boolean
    ): Long {
        val handlerId = nextHandlerId++

        var settings = AbstractBlock.Settings.create()
            .strength(hardness, resistance)

        if (requiresTool) {
            settings = settings.requiresTool()
        }

        val block = DartBlockProxy(settings, handlerId)
        blocks[handlerId] = block

        println("[ProxyRegistry] Created block with handler ID: $handlerId (hardness=$hardness, resistance=$resistance)")
        return handlerId
    }

    /**
     * Register the block with Minecraft's registry.
     * Must be called during mod initialization before registry freeze.
     *
     * Called from Dart via JNI.
     *
     * @param handlerId The handler ID of the block to register
     * @param namespace The namespace for the block ID (e.g., "mymod")
     * @param path The path for the block ID (e.g., "custom_block")
     * @return true if registration succeeded, false if block not found
     */
    @JvmStatic
    fun registerBlock(handlerId: Long, namespace: String, path: String): Boolean {
        val block = blocks[handlerId]
        if (block == null) {
            System.err.println("[ProxyRegistry] Cannot register block: handler ID $handlerId not found")
            return false
        }

        return try {
            val identifier = Identifier.of(namespace, path)

            // Register the block
            Registry.register(Registries.BLOCK, identifier, block)

            // Also register a BlockItem so it appears in creative inventory
            val blockItem = BlockItem(block, Item.Settings())
            Registry.register(Registries.ITEM, identifier, blockItem)

            println("[ProxyRegistry] Registered block: $identifier (handler ID: $handlerId)")
            true
        } catch (e: Exception) {
            System.err.println("[ProxyRegistry] Failed to register block $namespace:$path: ${e.message}")
            false
        }
    }

    /**
     * Get a block by its handler ID.
     *
     * @param handlerId The handler ID of the block
     * @return The block, or null if not found
     */
    @JvmStatic
    fun getBlock(handlerId: Long): DartBlockProxy? = blocks[handlerId]

    /**
     * Get all registered handler IDs.
     *
     * @return Array of all handler IDs
     */
    @JvmStatic
    fun getAllHandlerIds(): LongArray = blocks.keys.toLongArray()

    /**
     * Get the total number of registered blocks.
     *
     * @return The number of blocks in the registry
     */
    @JvmStatic
    fun getBlockCount(): Int = blocks.size

    /**
     * Check if a block with the given handler ID exists.
     *
     * @param handlerId The handler ID to check
     * @return true if the block exists
     */
    @JvmStatic
    fun hasBlock(handlerId: Long): Boolean = blocks.containsKey(handlerId)

    /**
     * Remove a block from the internal registry.
     * Note: This does NOT unregister it from Minecraft's registry (which is immutable after game start).
     *
     * @param handlerId The handler ID of the block to remove
     * @return true if the block was removed
     */
    @JvmStatic
    fun removeBlock(handlerId: Long): Boolean {
        val removed = blocks.remove(handlerId) != null
        if (removed) {
            println("[ProxyRegistry] Removed block with handler ID: $handlerId from internal registry")
        }
        return removed
    }
}
